up-l3/udiscovery/v3/service.adoc

```diff
@@ -0,0 +1,254 @@
+
+[.specitem,oft-sid="dsn~discovery-replication-client~1",oft-needs="impl"]
+--
+* link:../languages.adoc[uProtocol Language Libraries] *MUST NOT* implement the client-side of {replicator-proto-ref} API as these APIs are only used internally by uDiscovery business logic.
+--
+
+To replicate new or updated `UServiceTopic` metadata:
+[.specitem,oft-sid="dsn~discovery-setservicetopic-update~1",oft-needs="impl,test"]
+--
+* `ttl` value in `SetServiceTopicsRequest` *MUST* be set to the duration the metadata is valid for or absent from the message to indicate the metadata is valid forever.
+--
+
+To replicate removal of `UServiceTopic` metadata:
+[.specitem,oft-sid="dsn~discovery-setservicetopic-remove~1",oft-needs="impl,test"]
+--
+* `ttl` value in `SetServiceTopicsRequest` *MUST* be set to 0.
+--
+
+`SetServiceTopics()` Failure Reasons:
+
+[.specitem,oft-sid="dsn~discovery-setservicetopic-error-invalid~1",oft-needs="impl,test"]
+--
+*MUST* return `UCode.INVALID_ARGUMENT` if the URI passed does not contain a valid ue_id or a FQURI.
+--
+  
+[.specitem,oft-sid="dsn~discovery-setservicetopic-error-permission~1",oft-needs="impl,test"]
+--
+* *MUST* return `UCode.PERMISSION_DENIED` if the caller is not uDiscovery service in the correct recursive order (i.e. local -> domain -> central).
+* *MAY* grant access to deployment specific uEs such as a the software manager that installs or removes uEs.
+--
+
+[.specitem,oft-sid="dsn~discovery-setservicetopic-error-notfound~1",oft-needs="impl,test"]
+--
+* *MUST* return `UCode.NOT_FOUND` if no topic matching the URI was found when the `UServiceTopic` is being removed (i.e. `ttl=0`).
+--
+
+
+
+.Replication API Design
+[#udiscovery-replication-design]
+[mermaid]
+ifdef::env-github[[source,mermaid]]
+----
+sequenceDiagram
+
+OTAClient ->> Local-UDiscovery: SetServiceTopics()
+Local-UDiscovery -->> OTAClient: SetServiceTopicsResponse
+
+Local-UDiscovery ->> Domain-UDiscovery: SetServiceTopics()
+Domain-UDiscovery -->> Local-UDiscovery: SetServiceTopicsResponse
+
+Domain-UDiscovery ->> Central-UDiscovery: SetServiceTopics()
+Central-UDiscovery -->> Domain-UDiscovery: SetServiceTopicsResponse
+----
+
+
+== Implementation Consideration
+
+It is possible the information in domain or central instances might become out of sync with what is stored in the other instances. In order to rectify this situation:
+
+[.specitem,oft-sid="dsn~discovery-data-reconciliation~1",oft-needs="impl,test"]
+--
+* Domain and central instances *MUST* ensure that the information within their databases is accurate and up to date and reconcile any differences periodically using the client and replication APIs described in this specification.
```

Comment:
is the replication always only happening from _root_ to _leaf_, i.e. it is always initiated by a node asking its parent, never the other way around? This would also mean that nodes are only required to know their parent, but not the other way around, which would be good FMPOV.

Reply:
In an ideal world perhaps yes but due to the temporal nature of the leaves, it could be possible that the parents are not aware of changes in the leaves and they need to query the leaf (ex. the vehicle device does a factory reset of SW settings and there is no connectivity for a period of time. The replication needs to work in both directions to handle corner cases.

Reply:
Couldn't we instead specify that child nodes **MUST** (re-)sync with their parent node whenever they start up?

Reply:
@sophokles73 Ok I changed this, please take a look.